### DISCUSSION TITLE
How do I get the IP address from a request? remoteAddress in Node.

### DISCUSSION THREAD

**Discussion by banjo** [honojs/hono#1439] [General]
I'm using Hono with the node adapter, I'm currently trying to implement a rate limiter, but for that I need the incoming IP of the request, which I cannot seem to get unless it is from a header. Is there something like [remoteAddress](https://nodejs.org/dist/latest/docs/api/http.html#requestsocket) in node for Hono?

**Comment by yusukebe** [comment]
Hi @banjo 

> Is there something like [remoteAddress](https://nodejs.org/dist/latest/docs/api/http.html#requestsocket) in node for Hono?

Unfortunately, no. This is because the Request object in the web standard API has no field to have a remote address. But I think it's helpful if the developer can get a remote address in Node.js.

For Cloudflare, it adds the custom properties in the Request object: https://developers.cloudflare.com/workers/runtime-apis/request/#incomingrequestcfproperties

While this is not including an IP address, we might refer to the approach.

Even though if we implement it or not, keep this issue open. Thanks!

**Reply by icecake0107** [reply]
@yusukebe May I know how to get the original Cloudflare Request object in Hono? So I can read the `cf` properties?

**Reply by alex8bitw** [reply]
You can do:
c.req.header('CF-Connecting-IP')   # or True-Client-IP header.

If you want to get the request object you can do:

let newreq = c.req.raw.clone()

**Comment by yusukebe** [comment]
cc: @tangye

Do you have any opinion about this?

**Reply by charnould** [reply]
Any progress on this topic? It could be useful for some use cases. 👍 

Using Bun, it's not possible to get IP using at the same time Hono... :-(
https://bun.sh/blog/bun-v1.0.4#implement-server-requestip

**Reply by Zerebokep** [reply]
I agree, this is crucial to secure worker scripts from non-whitelisted ip's. Hopefully this will get implemented soon.

**Reply by imertz** [reply]
If you use a reverse proxy like nginx, you can pass the ip to the headers with:
proxy_set_header  X-Real-IP $remote_addr;

**Reply by leognmotta** [reply]
I believe that is a relevant topic, any progress?

**Comment by leognmotta** [comment]
Any news?

**Comment by usualoma** [comment]
If you are using `@hono/node-server`, you can get the `remoteAddress` with the following code.

```ts
import { serve } from '@hono/node-server'
import type { HttpBindings } from '@hono/node-server'
import { Hono } from 'hono'

const app = new Hono<{ Bindings: HttpBindings }>()

app.get('/', (c) => 
  c.text(`Hello, ${c.env.incoming.socket.remoteAddress}`)
)

serve(app)
```

**Reply by leognmotta** [reply]
Thank you I will try it out today, not sure if I found it in the docs
